specs/OAuth/AuthManager.md

```diff
@@ -0,0 +1,891 @@
+    NotifyFailure(tokenFailure.Error(), tokenFailure.ErrorDescription());
+}
+```
+
+Refreshing an Access Token (grant type/'response_type' = "code")
+
+```c#
+TokenRequestParams tokenRequestParams = TokenRequestParams::CreateForRefreshToken(refreshToken);
+ClientAuthentication clientAuth = ClientAuthentication::CreateForBasicAuthorization(L"my_client_id",
+    L"my_client_secret");
+TokenRequestResult tokenRequestResult = co_await AuthManager::RequestTokenAsync(
+    Uri(L"https://my.server.com/oauth/token"), tokenRequestParams, clientAuth));
+if (TokenResponse tokenResponse = tokenRequestResult.Response())
+{
+    UpdateToken(tokenResponse.AccessToken(), tokenResponse.TokenType(), tokenResponse.ExpiresIn());
+
+    //Store new refresh token if present
+    if (String refreshToken = tokenResponse.RefreshToken(); !refreshToken.empty())
+    {
+        // ExpiresIn is zero when not present
+        DateTime expires = winrt::clock::now();
+        if (String expiresIn = tokenResponse.ExpiresIn(); expiresIn != 0)
+        {
+            expires += std::chrono::seconds(static_cast<int64_t>(expiresIn));
+        }
+        else
+        {
+            // Assume a duration of one hour
+            expires += std::chrono::hours(1);
+        }
+
+        //Schedule a refresh of the access token
+        myAppState.ScheduleRefreshAt(expires, refreshToken);
+    }
+}
+else
+{
+    TokenFailure tokenFailure = tokenRequestResult.Failure();
+    NotifyFailure(tokenFailure.Error(), tokenFailure.ErrorDescription());
+}
+```
+
+Performing an Implicit Request for a token (grant type/'response_type' = "token")
+
+ ```c#
+AuthRequestParams authRequestParams = AuthRequestParams::CreateForImplicitRequest(L"my_client_id",
+    Uri(L"my-app:/oauth-callback/"));
+authRequestParams.Scope(L"user:email user:birthday");
+
+AuthRequestResult authRequestResult = co_await AuthManager::InitiateAuthRequestAsync(
+    Uri(L"https://my.server.com/oauth/authorize"), authRequestParams);
+if (AuthResponse authResponse = authRequestResult.Response())
+{
+    //To obtain the access token
+    String accessToken = tokenResponse.AccessToken();
+
+    String tokenType = tokenResponse.TokenType();
+
+    // Use the access token for resources
+    DoRequestWithToken(accessToken, tokenType);
+}
+else
+{
+    AuthFailure authFailure = authRequestResult.Failure();
+    NotifyFailure(authFailure.Error(), authFailure.ErrorDescription());
+}
+```
+
+> Note: The authorization server MUST NOT issue a refresh token for implicit request.
+
+Completing an Authorization Request from a Protocol Activation
+
+```c#
+void App::OnActivated(const IActivatedEventArgs& args)
+{
+    if (args.Kind() == ActivationKind::Protocol)
+    {
+        auto protocolArgs = args.as<ProtocolActivatedEventArgs>();
+        if (AuthManager::CompleteAuthRequest(protocolArgs.Uri()))
+        {
+            TerminateCurrentProcess();
+        }
+
+        DisplayUnhandledMessageToUser();
+    }
```

Comment:
I really don't like having to add this to all my apps, and force an app-shutdown. This is what WinUIEx does because it's the best it can do, but it's ugly and really should be handled at the app-platform before the entire application is started up (ie in main)

Reply:
If I understand correctly, you mean we should not terminate the instance at the app level, but this should be handled at the CompleteAuthRequest API level?

Reply:
If you keep it this way, put this into the "main" wherever that is - don't force the full App to load. It'll require adding code to that main, and for WinUI3 apps that means turning off the "XAML generated main" which is fine anyhow.  Then the "you are being launched to complete an OAuth2" can run just enough code in the main to complete the auth work and then exit (rather than terminate) the winmain without doing anything.

But maybe another way is the URI handler be part of the WASDK "main" package, and use a consistent URI scheme based on package identity.  The "main" package registers a protocol handler for "ms-wasdk-oauth" that goes to its very small "oauth2handler.exe.

When app-A (let's say it's packaged) wants to use OAuth2 here, the calling side forms the callback URI as "ms-wasdk-oauth:appname_publisherid/" ... when the site says `window.open(ms-wasdk-oauth:appnameA_publisheridA/goop/goes/here)` Windows will launch `oauth2handler.exe ms-wasdk-oauth:appnameA_publisheridA/goop/goes/here` in the WASDK main package. For app-B (also packaged) it goes to `ms-wasdk-oauth:appnameB_publisheridB/different/goop`. Both app-A and app-B register their handler URIs with their OAuth2 back-end as `ms-wasdk-oauth:appnameA_publisheridA/` and `ms-wasdk-oauth:appnameB_publisheridB` as appropriate.

The oauth2handler.exe takes the incoming request and cocreates some object in appA / appB and forwards the request onwards to the right object to complete.

Reply:
This universal handler approach sounds great. As per our discussion, we will aim for it for v2 post Experimental 1.7.